exercises/binary-search-tree/canonical-data.json

```diff
@@ -0,0 +1,68 @@
+{
+    "exercise": "binary-search-tree",
+    "version": "1.0.0",
+    "comments": [
+        "For simplicity only positive integer is considered as valid data,",
+        "null should be replaced by the respective language equivalent or -1",
+        "if the language does not have any.",
+
+        "The purpose of ths exercise is to create a data structure which",
+        "makes sure the data is always sorted, hence the exercise should",
+        "not make use of any sort() functions or sorting algorithms."
+    ],
+    "cases": [{
+            "description": "insert invalid/null data into the tree",
+            "property": "insert",
+            "value": null,
```

Comment:
Similar reasoning as #925: I don't think it is good to have this case.

> In most languages that have static type checking, even if they have nullable references, integers are not reference types, so if the function under test is defined to take an integer, `null` is not a possible input.
> 
> For languages without static type checking, `null` would be a possible input, but I also generally do not see functions that expect to take integers explicitly check for `null` (or language equivalent); let me know if I'm wrong. So I don't think the tests need to make students write it.

Reply:
As I stated in the comments : if the specific track does not have any equivalent  to `null` they should use `-1`. The purpose of this case is not limited to `null` but actually invalid data. I knew about #925 when I was writing this case and that's why I added the above comment.

Reply:
> if the specific track does not have any equivalent to null they should use -1. The purpose of this case is not limited to null but actually `invalid` data

I don't understand why -1 is invalid for any reason other than "the comments of this file say so".

I think it is a strange restriction to only allow positive numbers. It seems like the restriction is there for the sole purpose of classifying some numbers to be invalid, allowing there to be a test for inserting invalid values. If that is the sole purpose of the restriction, I would think it just makes more sense to not have the restriction and not have this test.

Further, if negative values are to be deemed invalid values for the tree being tested, that seems like information that is very important to be added in description.md, where students will see it, not canonical-data.json (where students will not see it!).

Reply:
Well the exercise is supposed to teach how to implement a data structure. And a data structure should know the difference between valid and invalid data before inserting it. Now some languages do not have static type checking whereas some don't have equivalent of null.  But we still need to draw a line between two. And I agree the description.md might need an update.

Reply:
I think this case should exist, because if we draw a line between valid and invalid data and specify in the description we should also have a test case to check against it.

Reply:
> And a data structure should know the difference between valid and invalid data before inserting it.

I expect that for a BST, by contract, the criterion for validity is a standard "are two elements comparable against each other?". I therefore also expect that this is the criterion that the tests will use, rather than a different standard ("is the number positive?", etc.)

So then, in a language with static type checking, this would get enforced by the type system. In a language without one, now it gets interesting. We could consider a test of adding in "hello" to a tree that already contains a 6, for example. This would fail. However, is it the student's responsibility to write the code that checks for that? The failure might happen without the student having to write any additional code. If it would, then there is no value in having this test. 

> the exercise is supposed to teach how to implement a data structure.

This makes it useful to keep the standard criterion for validity. If the exercise adds in an additional criterion, students may wonder why it exists and/or whether they have to add such an additional restriction in any future BSTs they write.

Reply:
So I guess the case for invalid data should exist but anything other than an integer should be considered as invalid.

Reply:
> anything other than an integer should be considered as invalid

Right.

> the case for invalid data should exist

Only if it is useful. I haven't been convinced it is useful because in most cases the student doesn't need to write any code to pass the test. Okay, I can name at least one language where `6 < "hello"` is not a type error (whether that be at run time or compile time), so in that language it could be useful.

I can't tell whether https://github.com/exercism/problem-specifications/issues/902 is telling us "yes, put such cases in the canonical-data.json and let languages that don't need the test ignore it" or "no, don't put such cases in and let languages that need the tests add them". I am thinking the number of languages that benefit from the case is lower than the number of languages that don't, if that helps decide.

I still say no. Remember that it's not necessary to convince me, it's only necessary to convince those who approve the pull request.

Reply:
A BST is about comperability, and I do not see why string and integer should not be comparable to each other...

In the erlang VM (and therefore in at least 3 languages on exercism) there is a total order defined over all values possible. `6 =< "foo"` is true in erlang, lfe and elixir (in their corresponding syntaxes).

Also after having read the introductionary comment, where you state that we only use positiv numbers as valid input, I'd use an unsigned int in statically typed languages.

But you have specified `-1` as an error indicator, but the only thing that can cause an error, is invalid input (which can't happen in accordingly typed languages) or someone else mangles with the tree from the outside, which we suppose not to happen.

So independently how I look at it, I do not feel that this test is necessary in any way.